src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs

```diff
@@ -225,16 +226,122 @@ public override bool DirectoryExists(string fullPath)
+            {
+                yield return storageFiles[i].Path;
+            }
+        }
+
+        private async static Task<IReadOnlyList<IStorageItem>> EnumerateFileQuery(string path, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
+        {
+            // Get a StorageFolder for "path"
+            string fullPath = Path.GetFullPath(path);
+            StorageFolder folder;
+            try
+            {
+                folder = await StorageFolder.GetFolderFromPathAsync(fullPath).TranslateWinRTTask(fullPath, isDirectory: true);
+            }
+            catch (FileNotFoundException e)
+            {
+                // GetFolderFromPathAsync throws FileNotFoundException but win32 throws a DirectoryNotFoundException for this case.
+                throw new DirectoryNotFoundException(e.Message, e.InnerException);
+            }
+
+            // Construct a query for the search.
+            QueryOptions query = new QueryOptions();
+
+            // Translate SearchOption into FolderDepth
+            query.FolderDepth = searchOption == SearchOption.AllDirectories ? FolderDepth.Deep : FolderDepth.Shallow;
+
+            // Construct an AQS filter
+            string normalizedSearchPattern = PathHelpers.NormalizeSearchPattern(searchPattern);
+            if (normalizedSearchPattern.Length == 0)
+            {
+                // An empty searchPattern will return no results and requires no AQS parsing.
+                return new IStorageItem[0];
+            }
+            else
+            {
+                // Parse the query as an ItemPathDisplay filter.
+                string searchPath = PathHelpers.GetFullSearchString(fullPath, normalizedSearchPattern);
+                string aqs = "System.ItemPathDisplay:~\"" + searchPath + "\"";
+                query.ApplicationSearchFilter = aqs;
+
+                // If the filtered path is deeper than the given user path, we need to get a new folder for it.
+                // This occurs when someone does something like Enumerate("C:\first\second\", "C:\first\second\third\*").
+                // When AllDirectories is set this isn't an issue, but for TopDirectoryOnly we have to do some special work
+                // to make sure something is actually returned when the searchPattern is a subdirectory of the path.
+                // To do this, we attempt to get a new StorageFolder for the subdirectory and return an empty enumerable
+                // if we can't.
+                string searchPatternDirName = Path.GetDirectoryName(normalizedSearchPattern);
+                string userPath = string.IsNullOrEmpty(searchPatternDirName) ? fullPath : Path.Combine(fullPath, searchPatternDirName);
+                if (userPath != folder.Path)
+                {
+                    try
+                    {
+                        folder = await StorageFolder.GetFolderFromPathAsync(userPath).TranslateWinRTTask(userPath, isDirectory: true);
```

Comment:
That is by design. The first call asserts that the initial folder exists and throws a DirectoryNotFoundException if it doesn't. In the case of a missing watcher path, it is not throw away work.

Cases I test:

path | searchPattern | result
------------ | ----------------- | --------------------------
```c:\temp\doesntexist``` | ```c:\temp\doesntexist\reallydoesntexist\*``` | throws DirectoryNotFoundException on first query
```c:\temp\doesntexist``` | ```c:\temp\doesntexist\*``` | throws DirectoryNotFoundException on first query
```c:\temp\exists``` | ```c:\temp\exists\subdirexists\*``` | does two queries - returns actual results
```c:\temp\exists``` | ```c:\temp\exists\subdirdoesntexist\*``` | does two queries - returns empty array

We can consolidate the calls, but it will require that we either:
- diverge from the win32 behavior and not throw a directorynotfoundexception on the first case in the table above
- Do a Directory.Exists call on the path when we have a subdirectory filter.

I don't want to diverge behavior (even for this edge case), so we need to do some sort of directory.exists call. The reason I left the second GetFolderFromPathAsync call is because it seemed cleaner and more reliable than adding a call to API that may determine existence of a directory differently (e.g. Directory.Exists may handle links differently than GetFolderFromPathAsync). Then there's the question of whether Directory.Exists is even *that much* faster than GetFolderFromPathAsync. 

There's also the fact that for the invalid-subpath case (number 1 in the table above), we currently quit execution on the first query and avoid doing path normalization and validation; if we consolidated the calls then we would then we would need to do two calls (GetFolderFromPathAsync and DirectoryExists) whereas now we only need one. We'd also be  slowing down **all** cases where the path doesn't exist because we'd be doing unnecessary path normalization before getting the (missing/invalid) folder.

TL;DR: the duplicate call is for checking the existence of the path before including the searchPattern and normalizing.
.